web/src/server/api/routers/comments.ts

```diff
@@ -0,0 +1,242 @@
+          projectId: input.projectId,
+          scope: "comments:CUD",
+        });
+
+        const comment = await ctx.prisma.comment.findFirst({
+          where: {
+            id: input.id,
+            projectId: input.projectId,
+            objectId: input.objectId,
+            objectType: input.objectType,
+          },
+        });
+        if (!comment) {
+          throw new Error("No comment with this id in this project.");
+        }
+
+        if (comment.authorUserId !== ctx.session.user.id) {
+          throw new Error(
+            "Comment author user id does not match provided user id",
+          );
+        }
+
+        await ctx.prisma.comment.delete({
+          where: {
+            id: comment.id,
+            projectId: comment.projectId,
+            objectId: comment.objectId,
+            objectType: comment.objectType,
+          },
+        });
+
+        await auditLog({
+          session: ctx.session,
+          resourceType: "comment",
+          resourceId: comment.id,
+          action: "delete",
+          before: comment,
+        });
+      } catch (error) {
+        console.error(error);
+        throw new TRPCError({
+          code: "INTERNAL_SERVER_ERROR",
+          message: "Deleting comment failed.",
+        });
+      }
+    }),
+  getByObjectId: protectedProjectProcedure
+    .input(
+      z.object({
+        projectId: z.string(),
+        objectId: z.string(),
+        objectType: z.nativeEnum(CommentObjectType),
+      }),
+    )
+    .query(async ({ input, ctx }) => {
+      try {
+        throwIfNoProjectAccess({
+          session: ctx.session,
+          projectId: input.projectId,
+          scope: "comments:read",
+        });
+
+        const comments = await ctx.prisma.$queryRaw<
+          Array<{
+            id: string;
+            content: string;
+            timestamp: Date;
+            authorUserId: string | null;
+            authorUserImage: string | null;
+            authorUserName: string | null;
+          }>
+        >(
+          Prisma.sql`
+        SELECT
+          c.id, 
+          c.content, 
+          c.timestamp,
+          u.id AS "authorUserId",
+          u.image AS "authorUserImage", 
+          u.name AS "authorUserName"
+        FROM comments c
+        LEFT JOIN users u ON u.id = c.author_user_id AND u.id in (SELECT user_id FROM organization_memberships WHERE org_id = ${ctx.session.orgId})
+        WHERE 
+          c."project_id" = ${input.projectId}
+          AND c."object_id" = ${input.objectId}
+          AND c."object_type"::text = ${input.objectType}
+        ORDER BY 
+          timestamp DESC
+        `,
+        );
+
+        return comments;
+      } catch (error) {
+        console.error(error);
+        throw new TRPCError({
+          code: "INTERNAL_SERVER_ERROR",
+          message: "Fetching comments by object id failed.",
+        });
+      }
+    }),
+  getCountsByObjectIds: protectedProjectProcedure
```

Comment:
allowing for multiple types seems like bad practice if there are collisions on ids across object types, should not be mixed but rather then done in separate trpc calls (e.g. one for trace and one for observations)